### PR TITLE
Kafka 0.10 APIs + Structured streaming test framework

### PR DESCRIPTION
**Pull request by ymahajan**
## Changes proposed in this pull request

- Removed Kafka 0.8 APIs from the product. 
- Upgraded product/test code to use Kafka 0.10 APIs   
- Retained both old SchemaDStream/CQ and new structured streaming
- Both streaming modules are based on Kafka 0.10 APIs now
- Fixed an issue reported here https://github.com/SnappyDataInc/snappydata/issues/862
- Cleaned up old unit tests, added more unit tests

## Patch testing

precheckin 

## ReleaseNotes.txt changes

Yes

## Other PRs
 
AQP - https://github.com/SnappyDataInc/snappy-aqp/pull/165

### CODE REVIEW

**Comment by sumwale** [comment]
@ymahajan Before reviewing, can we discuss a bit on the implications for users? Are there users using Kafka 0.8 with old streaming model with Spark 2.x? Would it not be better to retain both the models for sometime because Spark still has both. Have separate modules for the two like in Spark.
@jramnara

**Comment by ymahajan** [comment]
@sumwale @jramnara I don't see much value in maintaining three different streaming implementations. Kafka 0.8 is quite old now, even spark's structured streaming doesn't use it. It has been only implemented for Kafka 0.10. Spark's old streaming has separate external modules for 0.8 and 0.10. Combining them for us in product has it's own challenges in bundling, maintaining and adding test cases for both versions for old streaming, which anyways will be removed from the product eventually.

**Comment by jramnara** [comment]
Less complex the better. Do we have any stats on which version customers are using most? I suppose, 0.8 kafka is sufficient. Do some google searches.

**Comment by ymahajan** [comment]
@jramnara , Based on looking at stackoverflow and other forums, most of the kafka users are using 0.10, 0.11, 1.0 etc. And this 0.10 release is about to become almost 2 years stale now. Even structured streaming doesn't support anything less than 0.10 version. Also using 0.8 with spark streaming we needed to support both kafka and direct kafka streaming receiver. 

@rishitesh Jags wants to keep both streaming engine support for now. May be when we release structured streaming we will deprecate the legacy SchemaDStream, but for now we cannot remove it altogether. Also the legacy streaming supports features like stream to stream join which are not yet implemented in structured streaming.

File: core/src/main/scala/org/apache/spark/sql/streaming/DirectKafkaStreamSource.scala [hunk 1]
```diff
@@ -48,31 +53,112 @@ final class DirectKafkaStreamRelation(
...
+        })
+      }
+    stream
+  }
+}
+
+/**
+  * Copied from org.apache.spark.sql.kafka010.JsonUtils
+  */
+private object JsonUtils {
```
Comment (rishitesh) [inline core/src/main/scala/org/apache/spark/sql/streaming/DirectKafkaStreamSource.scala]: JsonUtil can be a separate class. Only saw usage in test classes. If so we can mode it to test packages.

**Pull request closed without merging**